lib/internal/modules/esm/loader.js

```diff
@@ -567,6 +568,8 @@ class ESMLoader {
       );
     }
 
+    new URL(url); // Intentionally trigger error if `url` is invalid
```

Comment:
My main concern would just be that there isn't a negligible performance overhead to these calls so if we know in advance that the URL is a valid URL (eg as we do for the default resolver), then it would be nice to avoid the double call as a performance optimization.

Reply:
Could perhaps it accept a string or a URL object? Then the string indicates it needs validation; the object indicates it doesn’t?

Reply:
I was going to propose as part of loaders chaining to change `url` to be a URL instance instead of an `URL::href`. My rationale being it saves the constant URL → href → URL → href currently in node core, and also necessary in custom loaders (so the argument of the `load` hook would be changed from a string to a URL instance, which we've already created within the ESMLoader, so then it just would not convert the existing URL instance to a string).

I think for the time-being, this extra URL construction is a drop in the pond, and the validation it provides is worth the cost: In discovering this issue, it took me a couple hours to track down a bad URL, and this instantly catches it.

Reply:
@JakobJingleheimer performance in Node.js is exactly a problem of a thousand cuts, so I wouldn't dismiss the concern entirely. I tend to agree a URL instance does make more sense despite the unfollowed whatwg guidance.

Reply:
Sorry, which unfollowed whatgw guidance?

Reply:
The alternative I can think of is regex, but I expect that to be worse for performance (I would guess under the hood `URL()` already uses an optimised and pre-compiled one), not to mention more error-prone.

Reply:
The WhatWG URL spec includes a note that specifically mentions not to use URL objects in APIs. I don't think we need to follow this guidance though necessarily as there are these perf and validation benefits to using URL objects.

All the usage of `new URL` in the resolver would likely be a target for performance optimization in avoiding reparsing or inlining just the relative normalization steps needed. Resolver performance as a whole is important though, as it does start to get felt as module counts grow.

There's no regex internally - just this https://github.com/nodejs/node/blob/master/src/node_url.cc#L822. It could even be faster to replace with a spec compliant regex guard (if a good subset can be found).

Reply:
Ah cool. Thanks!

Reply:
btw is that a 👍 from you, Guy?